### PR TITLE
Clear references to user tensors after compilation is finished

### LINKED ISSUE #2033: 3% Memory Regression when Compiling the Optimizer
### 🐛 Describe the bug

[Compiling the optimizer](https://github.com/pytorch/pytorch/pull/90709) introduced a ~3% memory regression as observed in the performance dashboard. @mlazos, @ngimel and others have tried debugging a bit but we haven't yet found the culprit.

3% is not huge but it's still worth getting to the bottom of this, and memory overheads add up.

### Error logs

_No response_

### Minified repro

_No response_

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/torchdynamo/issues/2033 
and https://github.com/pytorch/pytorch/issues/93490

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/convert_frame.py [hunk 1]
```diff
@@ -456,6 +456,7 @@ def transform(instructions, code_options):
         if hooks.guard_export_fn is not None:
             hooks.guard_export_fn(output.guards)
 
+        output.clear_live_user_objs()
```
Comment (eellison) [inline torch/_dynamo/convert_frame.py:459]: is there a reason this cant be here https://github.com/pytorch/pytorch/blob/40d6f2a02027023216607adb892d3b9c7493904c/torch/_dynamo/output_graph.py#L738 ?

**Review by eellison — COMMENTED** [review]
deferring review to @voznesenskym

File: torch/_dynamo/convert_frame.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/convert_frame.py]: Good point, done. Thought it was only cleaned once the code was invalidated by the CleanupManager

**Review by eellison — APPROVED** [review]
lgtm if tests pass

**Pull request closed without merging**